### ISSUE TITLE
Memory leak

### ISSUE THREAD

**Issue by canary01** [fadoss/maude-bindings#3]
I encountered a memory leak issue as in the following example.

```
import maude

maude.init(advise=False)
for i in range(1, 1000000):
    mod_def = f'''fmod A is pr NAT . eq s_^{i}(N:Nat) = N:Nat . endfm'''
    maude.input(mod_def)
    mod: maude.Module = maude.getModule('A')
    term = mod.parseTerm(f'{i}')
    term.reduce()
    if i % 10000 == 0:
        print(i)
```

Can I free the module object after I'm done with it?

**Comment by ningit** [comment]
In principle, modules should be garbage collected when they are no longer accessible, like in this case. However, due to some missing SWIG instructions, the Python `Module` object returned by `getModule` do not own the wrapped C++ Maude module, and its [destructor](https://github.com/fadoss/maude-bindings/blob/35c6875ba/swig/module.i#L34-L40) is not called when the Python object goes out of scope.

This bug will be fixed soon in the next release. Until then, a temporary solution would be manually setting `mod.thisown = True`. This can be done systematically by wrapping the `getModule` family of functions as below:

```python
def claimOwnership(fn):
    """Claim ownership of the object returned by fn"""
    def fn_own(*args, **kwargs):
        obj = fn(*args, **kwargs)
        obj.thisown = True
        return obj
    return fn_own

maude.getModule = claimOwnership(maude.getModule)
maude.getCurrentModule = claimOwnership(maude.getCurrentModule)
maude.downModule = claimOwnership(maude.downModule)
```

PD: if a different name is set for each module, like in `f'fmod A{i} is pr NAT . eq s_^{i}(N:Nat) = N:Nat . endfm'`, none will be released since they are still accessible by name.

**Comment by canary01** [comment]
I'm sorry for my late reply.

I have still 2 problems, but I've worked around them and am in no rush to fix it.
1. symbol.getModule still causes memory leak.
2. Following example occurs sigsegv because a module is collected even if it is referred from a term.

```
import maude


def claimOwnership(fn):
    """Claim ownership of the object returned by fn"""

    def fn_own(*args, **kwargs):
        obj = fn(*args, **kwargs)
        obj.thisown = True
        return obj

    return fn_own


maude.getModule = claimOwnership(maude.getModule)
maude.getCurrentModule = claimOwnership(maude.getCurrentModule)
maude.downModule = claimOwnership(maude.downModule)

maude.init(advise=False)

terms = []
for i in range(1, 1000000):
    name = 'A'
    mod_def = f'''fmod {name} is pr NAT . eq s_^{i}(N:Nat) = N:Nat . endfm'''
    maude.input(mod_def)
    mod: maude.Module = maude.getModule(name)
    term = mod.parseTerm(f'{i}')
    term.reduce()
    terms.append(term)
    if i % 10000 == 0:
        print(i, len(terms))
```

**Comment by ningit** [comment]
Thanks for reporting this bug, and sorry for the delay.

There is a fixed new version in the repository (1.2.1). Wheel packages are available [here](https://github.com/fadoss/maude-bindings/releases/tag/v1.2.1), although I have not uploaded them to PyPI yet. 

> symbol.getModule still causes memory leak.

This is already solved in the new version (a missing `%newobject` directive in SWIG). The temporary workaround of the `claimOwnership` function would also work for this case in the previous releases.

> Following example occurs sigsegv because a module is collected even if it is referred from a term.

Yes, the lifetime of modules was not properly handled, but the previous memory leaks and the fact that modules with a distinct name are not garbage collected masked the problem. This is a big issue since every object in the library (sorts, symbols, terms, equations, strategies, solution iterators, etc.) belongs to a module, which must survive them. In the new version, the constructors and destructors of these objects update the reference count of the internal Maude module so that it is released only when appropriate. However, some corner (somehow artificial) cases are not covered to avoid adding additional layers of indirection (the [module-life.py](https://github.com/fadoss/maude-bindings/blob/07b4cf7c38c31b46176c124032e0a2f4bdf051bf/tests/python/module-life.py) test includes some examples). In particular, terms always ensure that the module is not garbage collected, so your example works well.

A workaround for the previous versions when used with `claimOwnership` (and for the corner cases of the new version, although they seem unlikely to occur in practice) is keeping the `Module` object the entire time its elements are being used. This (without the extensive reference counting) would be the most efficient approach (for example, now `list(m.getSymbols())` for `CONVERSION` does 202 additional operations), but that requirement in a garbage-collected language like Python does not seem reasonable.

**Comment by canary01** [comment]
I couldn't get rid of the segmentation fault with the patched version, but it worked fine with version 1.2.1.
Sincerely thank.

**Comment by ningit** [comment]
There was a bug introduced in 1.2.1 (when `findSort` does not find the sort), but it is now fixed in 1.2.2. Everything else seems to work fine, so I have uploaded the new version to PyPI. Thanks again.

> I couldn't get rid of the segmentation fault with the patched version

Yes, the `claimOwnership` patch only prevents the memory leak.
